cmd/root.go

```diff
@@ -76,23 +86,112 @@ func newClient(domain string, scopes ...string) *client.Client {
+	var err error
+	useHTTPS := false
+	if envHTTPS := os.Getenv("COZY_ADMIN_HTTPS_CLIENT"); envHTTPS != "" {
+		useHTTPS, err = strconv.ParseBool(envHTTPS)
+		if err != nil {
+			errFatalf("Could not read COZY_ADMIN_HTTPS variable: %s", err)
+		}
+	}
 	pass := []byte(os.Getenv("COZY_ADMIN_PASSWORD"))
 	if !config.IsDevRelease() {
 		if len(pass) == 0 {
 			var err error
 			fmt.Printf("Password:")
 			pass, err = gopass.GetPasswdMasked()
 			if err != nil {
-				errPrintf("Could not get password from standard input: %s\n", err)
-				os.Exit(1)
+				errFatalf("Could not get password from standard input: %s\n", err)
 			}
 		}
 	}
-	return &client.Client{
+	c := &client.Client{
 		Domain:     config.AdminServerAddr(),
-		Scheme:     "http",
 		Authorizer: &request.BasicAuthorizer{Password: string(pass)},
 	}
+	if useHTTPS {
+		c.Scheme = "https"
+		c.Client = sslClient()
+	} else {
+		c.Scheme = "http"
+	}
+	return c
+}
+
+func sslClient() *http.Client {
+	var rootCAs *x509.CertPool
+	var clientCertificate tls.Certificate
+	var verifyPeerCertificate func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error
+
+	if envRootCA := os.Getenv("COZY_ADMIN_HTTPS_CLIENT_ROOTCA_FILE"); envRootCA != "" {
+		rootCA, err := ioutil.ReadFile(envRootCA)
+		if err != nil {
+			errFatalf("Could not read file %q: %s", envRootCA, err)
+		}
+		rootCAs = x509.NewCertPool()
+		rootCAs.AppendCertsFromPEM(rootCA)
+	}
+
+	if envClientCert := os.Getenv("COZY_ADMIN_HTTPS_CLIENT_CERT_FILE"); envClientCert != "" {
+		envClientKeyFile := os.Getenv("COZY_ADMIN_HTTPS_CLIENT_KEY_FILE")
+		cert, err := tls.LoadX509KeyPair(envClientCert, envClientKeyFile)
+		if err != nil {
+			errFatalf("Could not read client certificate files %q and %q: %s",
+				envClientCert, envClientKeyFile, err)
+		}
+		clientCertificate = cert
+	}
+
+	if envKeyPinned := os.Getenv("COZY_ADMIN_HTTPS_CLIENT_KEYPINNED_FINGERPRINT"); envKeyPinned != "" {
+		pinnedFingerPrint, err := hex.DecodeString(envKeyPinned)
+		if err != nil {
+			pinnedFingerPrint, err = base64.StdEncoding.DecodeString(envKeyPinned)
+		}
+		if err != nil {
+			errFatalf("Invalid encoding for COZY_ADMIN_HTTPS_CLIENT_KEYPINNED_FINGERPRINT")
+		}
+		if len(pinnedFingerPrint) != sha256.Size {
+			errFatalf("Invalid size for COZY_ADMIN_HTTPS_CLIENT_KEYPINNED: expected %d got %d",
+				sha256.Size, len(pinnedFingerPrint))
+		}
+		verifyPeerCertificate = sslVerifyPinnedKey(pinnedFingerPrint)
+	}
+
+	tlsConfig := &tls.Config{
+		Certificates:          []tls.Certificate{clientCertificate},
+		RootCAs:               rootCAs,
+		VerifyPeerCertificate: verifyPeerCertificate,
+		InsecureSkipVerify:    false, // should be false, we *need* rootca verification
+	}
+	return &http.Client{
+		Timeout:   15 * time.Second,
+		Transport: &http.Transport{TLSClientConfig: tlsConfig},
+	}
+}
+
+func sslVerifyPinnedKey(pinnedFingerPrint []byte) func(rawCerts [][]byte, _ [][]*x509.Certificate) error {
+	if len(pinnedFingerPrint) != sha256.Size {
```

Comment:
Already verified L153

Reply:
This is just me being extra-careful if we use this method in the future without this verification...